Packet++/header/InfiniBandLayer.h

```diff
@@ -0,0 +1,288 @@
+		 * @param[in] qpn The destination queue pair (QP) number
+		 * @param[in] ack_req The acknowledgment request
+		 * @param[in] psn The packet sequence number
+		 */
+		InfiniBandLayer(uint8_t opcode, int se, int mig, int pad, uint16_t pkey, uint32_t qpn, int ack_req,
+		                uint32_t psn);
+
+		/**
+		 * Get a pointer to the BTH header. Notice this points directly to the data, so every change will change
+		 * the actual packet data
+		 * @return A pointer to the bth_header
+		 */
+		rxe_bth* getBthHeader() const
+		{
+			return reinterpret_cast<rxe_bth*>(m_Data);
+		}
+
+		/**
+		 * @return The operation code which defines the interpretation of the remaining header and payload bytes
+		 */
+		uint8_t getOpcode() const;
+
+		/**
+		 * Set operation code
+		 * @param[in] opcode The opcode to set
+		 */
+		void setOpcode(uint8_t opcode) const;
+
+		/**
+		 * @return solicited event that the responder shall invoke the CQ event handler
+		 */
+		uint8_t getSe() const;
+
+		/**
+		 * Set solicited event
+		 * @param[in] se The solicited event to set
+		 */
+		void setSe(int se) const;
+
+		/**
+		 * @return migreq which used to communicate migration state
+		 */
+		uint8_t getMig() const;
+
+		/**
+		 * Set migreq
+		 * @param[in] mig The migration state to set. If set to one, indicates the connection or EE context has been
+		 * migrated; if set to zero, it means there is no change in the current migration state.
+		 */
+		void setMig(uint8_t mig) const;
+
+		/**
+		 * @return PadCount which Packet payloads are sent as a multiple of 4-byte quantities.
+		 * Pad count indicates the number of pad bytes - 0 to 3 - that are appended to the packetpayload.
+		 * Pads are used to “stretch” the payload (payloads may be zero or more bytes in length) to be a multiple of 4
+		 * bytes
+		 */
+		uint8_t getPad() const;
+
+		/**
+		 * Set PadCount
+		 * @param[in] pad The PadCount to set
+		 */
+		void setPad(uint8_t pad) const;
+
+		/**
+		 * @return Transport Header Version that specifies the version of the IBA Transport used for this packet
+		 */
+		uint8_t getTver() const;
+
+		/**
+		 * Set Transport Header Version
+		 * @param[in] tvr The transport header version to set
+		 */
+		void setTver(uint8_t tver) const;
+
+		/**
+		 * @return partition key identifying the partition
+		 * that the destination QP (RC, UC, UD, XRC) or EE Context (RD) is a member.
+		 */
+		uint16_t getPkey() const;
+
+		/**
+		 * Set partition key
+		 * @param[in] pkey The partition key to set
+		 */
+		void setPkey(uint16_t pkey) const;
+
+		/**
+		 * @return destination queue pair (QP) identifier
+		 */
+		uint32_t getQpn() const;
+
+		/**
+		 * Set Queue Pair Number
+		 * @param[in] qpn The queue pair number to set
+		 */
+		void setQpn(uint32_t qpn) const;
+
+		/**
+		 * @return FECN
+		 * F (FECN): 0 indicates that a FECN indication was not received.
+		 * 1 indicates that the packet went through a point of congestion
+		 */
+		int getFecn() const;
```

Comment:
FECN, BECN and Ack are all 1-bit flags, maybe they should be represented by `bool` instead of `int`?

Reply:
```
static inline int __bth_fecn(void *arg)
{
	struct rxe_bth *bth = arg;

	return 0 != (cpu_to_be32(BTH_FECN_MASK) & bth->qpn);
}

static inline void __bth_set_fecn(void *arg, int fecn)
{
	struct rxe_bth *bth = arg;

	if (fecn)
		bth->qpn |= cpu_to_be32(BTH_FECN_MASK);
	else
		bth->qpn &= ~cpu_to_be32(BTH_FECN_MASK);
}

static inline int __bth_becn(void *arg)
{
	struct rxe_bth *bth = arg;

	return 0 != (cpu_to_be32(BTH_BECN_MASK) & bth->qpn);
}
```
the linux kernel implement it with `int` instead of `bool`, maybe we can follow their design.

Reply:
As far as I know kernel code often uses `int` instead of `bool`, but PcapPlusPlus is higher level and I think `bool` is more explicit

Reply:
ok